cmd/earth/main.go

```diff
@@ -121,8 +144,7 @@ func (app *earthApp) autoComplete() {
 
 	potentials, err := autocomplete.GetPotentials(compLine, int(compPointInt), flags, commands)
 	if err != nil {
-		//panic(err) // can't display error or it will show up under tab-completion
-		os.Exit(1)
+		panic(err)
```

Comment:
I would avoid panic-recover and instead just return an err which can be logged in autocomplete.log. It'll make it a bit more consistent / easier to follow. We should generally reserve panic for completely unexpected situations.

Reply:
We need to have the recover code in place to catch a panic which occurs in the antlr library.

I figured I could take advantage of the recovery code for the case where getPotentials also returns an error rather than just silently ignoring it.

If we did move the code that opens autocomplete.log, and if there's an error opening up the file, then do 1) we silently ignore it and exit (thus breaking auto completion), or 2) leave the file pointer as nil and check that it's not nil before logging to it?

The other option could be something like:
```
func autoComplete() error {
    var err error
    defer func() {
        err = recover()
    }

    doStuff()

    return err
}
```

then under main we could log the error to autocomplete?

Reply:
I like your last option. It makes most sense to me. It pretty much keeps the panic recover isolated to antlr stuff.

Antlr Go API is such a mess :-)

Reply:
So I tried moving this around, and the one thing I don't really like is now that the exit code is in main which is a bit confusing:

```
func (app *earthApp) autoComplete() (bool, error) {
	_, found := os.LookupEnv("COMP_LINE")
	if !found {
		return false, nil
	}

	var err error
	defer func() {
		if r := recover(); r != nil {
			err = fmt.Errorf("recovered panic in autocomplete %v: %v", r, debug.Stack())
		}
	}()

	compLine := os.Getenv("COMP_LINE")   // full command line
	compPoint := os.Getenv("COMP_POINT") // where the cursor is

	compPointInt, err := strconv.ParseUint(compPoint, 10, 64)
	if err != nil {
		return err
	}

	flags := []string{}
	for _, f := range app.cliApp.Flags {
		for _, n := range f.Names() {
			if len(n) > 1 {
				flags = append(flags, n)
			}
		}
	}

	commands := []string{}
	for _, cmd := range app.cliApp.Commands {
		commands = append(commands, cmd.Name)
	}

	potentials, err := autocomplete.GetPotentials(compLine, int(compPointInt), flags, commands)
	if err != nil {
		return true, err
	}
	for _, p := range potentials {
		fmt.Printf("%s\n", p)
	}

	return true, err
}
```

then under main we need:

```
	shouldExit, err := app.autoComplete()
	if err != nil {
		homeDir, err := os.UserHomeDir()
		if err != nil {
			os.Exit(1)
		}
		logDir := filepath.Join(homeDir, ".earthly")
		logFile := filepath.Join(logDir, "autocomplete.log")
		err = os.MkdirAll(logDir, 0755)
		if err != nil {
			os.Exit(1)
		}
		f, err := os.OpenFile(logFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0755)
		if err != nil {
			os.Exit(1)
		}
		fmt.Fprintf(f, "error during autocomplete: %v\n", err)
		os.Exit(1)
	}
        if shouldExit {
            os.Exit(0)
        }
```

Reply:
instead we could have two functions: `autoComplete()` and `autoCompleteImp() error`:

```
func (app *earthApp) autoComplete() {
	_, found := os.LookupEnv("COMP_LINE")
	if !found {
		return
	}

	err := s.autoCompleteImp()
	if err != nil {
		homeDir, err := os.UserHomeDir()
		if err != nil {
			os.Exit(1)
		}
		logDir := filepath.Join(homeDir, ".earthly")
		logFile := filepath.Join(logDir, "autocomplete.log")
		err = os.MkdirAll(logDir, 0755)
		if err != nil {
			os.Exit(1)
		}
		f, err := os.OpenFile(logFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0755)
		if err != nil {
			os.Exit(1)
		}
		fmt.Fprintf(f, "error during autocomplete: %v\n", err)
		os.Exit(1)
	}
	os.Exit(0)
}

func (app *earthApp) autoCompleteImp() error {
	_, found := os.LookupEnv("COMP_LINE")
	if !found {
		return nil
	}

	var err error
	defer func() {
		if r := recover(); r != nil {
			err = fmt.Errorf("recovered panic in autocomplete %v: %v", r, debug.Stack())
		}
	}()

	compLine := os.Getenv("COMP_LINE")   // full command line
	compPoint := os.Getenv("COMP_POINT") // where the cursor is

	compPointInt, err := strconv.ParseUint(compPoint, 10, 64)
	if err != nil {
		return err
	}

	flags := []string{}
	for _, f := range app.cliApp.Flags {
		for _, n := range f.Names() {
			if len(n) > 1 {
				flags = append(flags, n)
			}
		}
	}

	commands := []string{}
	for _, cmd := range app.cliApp.Commands {
		commands = append(commands, cmd.Name)
	}

	potentials, err := autocomplete.GetPotentials(compLine, int(compPointInt), flags, commands)
	if err != nil {
		return err
	}
	for _, p := range potentials {
		fmt.Printf("%s\n", p)
	}

	os.Exit(0)
}
```

Reply:
We could extract

```
_, found := os.LookupEnv("COMP_LINE")
	if !found {
		return false, nil
	}
```

to main as well? It's the only thing influencing `shouldExit`.

Reply:
> instead we could have two functions: autoComplete() and autoCompleteImp() error

✔️

Reply:
I spoke too soon with my idea, apparently it won't work like that. Here's an example:

```
package main

import (
	"fmt"
)

func foo() error {
    var err error
    defer func() {
        if r := recover(); r != nil {
            fmt.Printf("will return err\n")
            err = fmt.Errorf("got err")
        }
    }()

    if 4 > 1 {
        panic("ugh")
    }

    fmt.Printf("will return nil\n")
    err = nil
    return err

}

func main() {
	fmt.Printf("got %v\n", foo())
}
```
try it out at https://play.golang.org/p/W8t832KAsl4

Reply:
nevermind, figured it out with https://play.golang.org/p/l_s0zBBYeZ9 ignore my last comment.